streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -3302,6 +3304,91 @@ object ZStream extends ZStreamPlatformSpecificConstructors with Serializable {
   def fromJavaIteratorManaged[R, E, A](iterator: ZManaged[R, E, ju.Iterator[A]]): ZStream[R, E, A] =
     managed(iterator).flatMap(StreamEffect.fromJavaIterator)
 
+  /**
+   * Creates a managed (impure) Iterator from a Stream
+   */
+  def toIterator[R, E, A](
+    stream: ZStream[R, E, A],
+    afterLastPull: URIO[R, Any] = ZIO.unit
+  ): ZManaged[R, Nothing, Iterator[Either[E, A]]] = {
+    val nextOnAnEmptyIterator = new NoSuchElementException("next on empty iterator")
+
+    /*
+     * Internal state of a Iterator pulling from a ZStream
+     *
+     * It starts as Running
+     *
+     * when Running  , on pull (hasNext), pull the ZStream and switch to Closed or Value
+     * when Value    , on consume (next), return the Value and switch to Running
+     * when Closed   , on pull (hasNext), stays Closed
+     */
+    sealed trait State
+    case object Running                   extends State
+    sealed trait ValueOrClosed            extends State
+    case object Closed                    extends ValueOrClosed
+    case class Value(value: Either[E, A]) extends ValueOrClosed
+
+    for {
+      pull    <- stream.process
+      runtime <- ZIO.runtime[R].toManaged_
+    } yield {
+      new Iterator[Either[E, A]] {
+
+        var state: State = Running
+
+        var executedFinalizer: Boolean = false
+
+        private def finalizeIterator(): Unit =
+          if (!executedFinalizer) {
+            executedFinalizer = true
+            runtime.unsafeRun(afterLastPull.unit)
+          }
+
+        private def pool(): ValueOrClosed =
+          runtime.unsafeRun(
+            pull
+              .fold({
+                case None    => Closed
+                case Some(e) => Value(Left(e))
+              }, x => Value(Right(x)))
+              .tap({
+                case Closed => ZIO.effectTotal(finalizeIterator())
+                case _      => ZIO.unit
+              })
+          )
+
+        @tailrec
+        override def hasNext: Boolean =
+          state match {
+            case Closed => false
+
+            //Should not happens, unless hasNext is called twice in a row.
```

Comment:
I think we have to assume that `hasNext` and `next` may be called in arbitrary sequences here.